Updating the touchscreen interface structure

nielsnl68 opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

**This breaking PR implements a consistent solution for the current available touchscreen drivers.**
- A `binary sensor` and `on_touch` event will now be triggered once until all touches are released. 
- Two new events are added:
   - `on_update` will be triggered when a touched position moved around or that an extra touch has been detected.
   - `on_release` will be triggered when all touches are release from the touchscreen.
- The raw touch position is added to the `on_touch` `touch` argument.
- The touchscreen drivers will from now on get the raw touch data on request of main touchscreen component.
  - the touchscreen component will convert the raw data and determine which triggers and events will be called.
 
**NOTE: This is a PR that will break all existing touchscreen components not in the esphome repository,** 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3405
## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
touchscreen:
  - platform: xpt2046
    id: touch_1
    display: my_display
    ...
  ```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nielsnl68 committed:
make Display buffer ID optional

nielsnl68 committed:
replace the width and height to get them directly

nielsnl68 committed:
clang fixes

nielsnl68 committed:
clang fixes

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -31,14 +34,21 @@ enum TouchRotation {
   display::DisplayBuffer *get_display() const { return this->display_; }
-
+#else
+  void set_dispay_dimension(uint16_t width, u_int16_t height) {
```
```suggestion
  void set_display_dimension(uint16_t width, u_int16_t height) {
```

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -47,10 +57,38 @@ class Touchscreen {
+#endif
+  }
+
+  uint16_t get_rotation_() {
```
Why not return the enum type?
 ```suggestion
  TouchRotation get_rotation_() {
```

nielsnl68 replied:
thanks, you are right, i missed this. I was to busy try it to work without the need of a display

nielsnl68 committed:
Update esphome/components/touchscreen/touchscreen.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
Update esphome/components/touchscreen/touchscreen.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
clean up and add property handling in __init__.py

nielsnl68 committed:
Add define USE_DISPLAY when a display is used.

nielsnl68 committed:
resolve some issue around USE_DISPLAY introduction

nielsnl68 committed:
resolve some issue around USE_DISPLAY introduction

nielsnl68 committed:
fix clang issue

nielsnl68 committed:
i missed a function call

nielsnl68 committed:
i missed a function call

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -47,10 +54,50 @@ class Touchscreen {
+#ifdef USE_DISPLAY
+  display::DisplayBuffer *display_{nullptr};
+#else
+  uint16_t display_width_{240};
+  uint16_t display_height_{320};
```
The config validation should enforce at least either `display` or `dimensions`+`rotation` to be supplied and remove these fallback values from this file.

nielsnl68 replied:
Good point.
I am try to figure out how i should do that correctly. I could not find a suitable function in `config_validation` to do what you suggest, i think.

nielsnl68 replied:
Added this check now

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
moving code around.

nielsnl68 committed:
fixing my mistakes  but one left

nielsnl68 committed:
update codeowners

nielsnl68 committed:
fixing last issue

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
update all touchscreen drivers.

nielsnl68 committed:
remove clang issues

nielsnl68 committed:
update codeowners

nielsnl68 committed:
fix one in test8

nielsnl68 committed:
remove USE_DISPLAY define from Display component

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
add new triggers on_update and on_release

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

nielsnl68 committed:
Update events parameter names and types
and hide display class when not in use.

nielsnl68 committed:
trying to fix test4 yaml issue

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
removed unused variable

nielsnl68 committed:
add extra check in validate_xpt2046

nielsnl68 committed:
update CONFIG_SCHEMA structures

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
restore touchscreen.py

nielsnl68 committed:
oeps wrong folder

nielsnl68 committed:
resolve upgrade conflict

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
update codeowners

nielsnl68 committed:
update codeowners

nielsnl68 committed:
restore xpt2048.cpp to old location

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
remove temporarily restored files

jesserockz commented on esphome/components/touchscreen/__init__.py:
```diff
@@ -3,7 +3,7 @@
+from esphome.const import CONF_ON_TOUCH, CONF_ON_RELEASE, CONF_DIMENSIONS, CONF_ROTATION
 from esphome.core import coroutine_with_priority
 
 CODEOWNERS = ["@jesserockz"]
```
Maybe add yourself here now that you are making many changes?

nielsnl68 replied:
done

nielsnl68 committed:
added myself as the contributor

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Remove the optional Display future for now.

nielsnl68 committed:
rework the way touches are registered.
and added previous and orginal position in the TouchPoint record/struct.
removed obsolute report_interval

nielsnl68 committed:
remove unneeded spaces

nielsnl68 committed:
update codeowners and spacing

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
fix method name

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
removed unneeded variable

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'esphome:dev' into nvds-touchscreen-optdisp

gaaf committed:
core: Don't delay loop if already late

The application loop would delay a full loop_interval_ when the last loop
run already took more than loop_interval_ time.

Instead, don't delay if we're already late and run the next loop as soon
as possible.

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,26 +1,43 @@
+  virtual void update(const TouchPoints_t &tpoints) {}
   virtual void release() {}
 };
+// std::vector<Style> styles_{};
```
Commented code?

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,26 +1,43 @@
-  virtual void touch(TouchPoint tp) = 0;
+  virtual void touch(TouchPoint tp) {}
+  virtual void touch(const TouchPoints_t &tpoints) {}
+  /* return true to update first touch position */
```
What does this comment refer to exactly?
There are no bool returns here?

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -29,25 +46,63 @@ enum TouchRotation {
+
+  uint16_t get_width_() { return this->display_->get_width(); }
+
+  uint16_t get_height_() { return display_->get_height(); }
```
```suggestion
  uint16_t get_height_() { return this->display_->get_height(); }
```

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
remove old and unrelated comments

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

jesserockz commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -29,25 +44,63 @@ enum TouchRotation {
+  TouchRotation get_rotation_() { return static_cast<TouchRotation>(this->display_->get_rotation()); }
+  display::Display *display_{nullptr};
+
+  int16_t x_raw_min_{0}, x_raw_max_{0}, y_raw_min_{0}, y_raw_max_{0};
```
Do these need to be signed values?

```suggestion
  uint16_t x_raw_min_{0}, x_raw_max_{0}, y_raw_min_{0}, y_raw_max_{0};
```

nielsnl68 replied:
No, i want to allow negative values as well.

jesserockz commented on esphome/components/touchscreen/touchscreen.cpp:
```diff
@@ -7,22 +7,130 @@ namespace touchscreen {
+int16_t Touchscreen::normalize(int16_t val, int16_t min_val, int16_t max_val, bool inverted) {
+  int16_t ret;
+
+  if (val <= min_val) {
+    ret = 0;
+  } else if (val >= max_val) {
+    ret = 0xfff;
+  } else {
+    ret = (int16_t) ((int) 0xfff * (val - min_val) / (max_val - min_val));
+  }
```
This could be replace by `clamp` and `remap` from `helpers.h` =)

nielsnl68 replied:
Good idea, need to dig in to this a bit more. Can i do that for an upcoming fix?

jesserockz commented on esphome/components/touchscreen/binary_sensor/touchscreen_binary_sensor.h:
```diff
@@ -1,7 +1,11 @@
+#include "esphome/core/defines.h"
 #include "esphome/components/binary_sensor/binary_sensor.h"
+
+#ifdef USE_DISPLAY
```
Since display is currently required by touchscreen, we can safely assume it is there.

jesserockz commented on esphome/components/touchscreen/binary_sensor/touchscreen_binary_sensor.cpp:
```diff
@@ -10,11 +10,11 @@ void TouchscreenBinarySensor::setup() {
   bool touched = (tp.x >= this->x_min_ && tp.x <= this->x_max_ && tp.y >= this->y_min_ && tp.y <= this->y_max_);
-
-  if (this->page_ != nullptr) {
+#ifdef USE_DISPLAY
```
Same here

jesserockz commented on esphome/components/touchscreen/__init__.py:
```diff
@@ -3,44 +3,84 @@
 TOUCHSCREEN_SCHEMA = cv.Schema(
     {
-        cv.GenerateID(CONF_DISPLAY): cv.use_id(display.DisplayBuffer),
+        cv.Required(CONF_DISPLAY): cv.use_id(display.DisplayBuffer),
```
Lets not break the config here until we are making changes for it to become optional
```suggestion
        cv.GenerateID(CONF_DISPLAY): cv.use_id(display.DisplayBuffer),
```

jesserockz commented on esphome/components/touchscreen/__init__.py:
```diff
@@ -3,44 +3,84 @@
+        cv.Required(CONF_DISPLAY): cv.use_id(display.DisplayBuffer),
+        cv.Optional(CONF_DIMENSIONS): cv.dimensions,
+        cv.Optional(CONF_ROTATION): validate_rotation,
+        cv.Optional(CONF_SWAP_X_Y, default=False): cv.boolean,
```
These are not used?

jesserockz commented on esphome/components/lilygo_t5_47/touchscreen/lilygo_t5_47_touchscreen.cpp:
```diff
@@ -69,102 +61,35 @@ void LilygoT547Touchscreen::loop() {
+    point = 1;
+
+  uint16_t id, x_raw, y_raw;
+  for (int i = 0; i < point; i++) {
```
```suggestion
  for (uint8_t i = 0; i < point; i++) {
```

jesserockz commented on esphome/components/lilygo_t5_47/touchscreen/lilygo_t5_47_touchscreen.cpp:
```diff
@@ -69,102 +61,35 @@ void LilygoT547Touchscreen::loop() {
+    // not sure what todo with this atm.
+    //    tp.state = buffer[i * 5] & 0x0F;
+    //    if (tp.state == 0x06)
+    //      tp.state = 0x07;
```
If its not used, lets just delete it
```suggestion
```

jesserockz commented on esphome/components/xpt2046/touchscreen/__init__.py:
```diff
@@ -0,0 +1,99 @@
+        raise cv.Invalid("Calibration X values missing")
+
+    if CONF_CALIBRATION_Y_MAX not in config or CONF_CALIBRATION_Y_MIN not in config:
+        raise cv.Invalid("Calibration y values missing")
```
They cant be missing since they have a default value

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,114 @@
+    this->attach_interrupt_(this->irq_pin_, gpio::INTERRUPT_FALLING_EDGE);
+  }
+  spi_setup();
+  read_adc_(0xD0);  // ADC powerdown, enable PENIRQ pin
```
I know you probably just moved this file for the most part, but it shows up as an entirely new file on github due to the changes, so it may as well be cleaned up to standards =)

```suggestion
  this->spi_setup();
  this->read_adc_(0xD0);  // ADC powerdown, enable PENIRQ pin
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,114 @@
+  ESP_LOGCONFIG(TAG, "  Invert Y: %s", YESNO(this->invert_y_));
+
+  ESP_LOGCONFIG(TAG, "  threshold: %d", this->threshold_);
+  // ESP_LOGCONFIG(TAG, "  Report interval: %u", this->report_millis_);
```
```suggestion
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,114 @@
+float XPT2046Component::get_setup_priority() const { return setup_priority::DATA; }
+
+int16_t XPT2046Component::best_two_avg(int16_t x, int16_t y, int16_t z) {  // NOLINT
+  int16_t da, db, dc;                                                      // NOLINT
```
Maybe name these `delta_a` etc instead of disabling lint =)

jesserockz commented on tests/test8.yaml:
```diff
@@ -65,6 +66,7 @@ i2c:
 
 touchscreen:
   - platform: tt21100
+    display: displ8
```
Should not be required if reverted to `GenerateID`

nielsnl68 committed:
Update esphome/components/touchscreen/touchscreen.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
Update esphome/components/xpt2046/touchscreen/xpt2046.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
Update esphome/components/xpt2046/touchscreen/xpt2046.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
impl Jesse's suggestions

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

nielsnl68 committed:
impl Jesse's suggestions #2

gpambrozio committed:
Adding touchscreen functionality for the FT63X6 driver (#266)

* Adding touchscreen functionality for the FT63X6 driver

* Apply suggestions from code review

* Apply suggestions from code review

* Apply suggestions from code review

* Update esphome/components/ft63x6/ft63x6.cpp

---------

Co-authored-by: NP v/d Spek <github_mail@lumensoft.nl>

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
update codeowners

nielsnl68 committed:
update gt911 touchscreen

clydebarrow commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -49,27 +47,22 @@ void GT911Touchscreen::setup() {
   this->write(CLEAR_TOUCH_STATE, sizeof(CLEAR_TOUCH_STATE));
 
-  if ((touch_state & 0x80) == 0)
+  if (touch_state  == 0)
```
```suggestion
  if ((touch_state & 0x80) == 0)
```

nielsnl68 replied:
@Clyde, why would you want that change back, In both case (touch_state & 0x80) and (touch_state & 0x07) would mean a return without changes.
Maybe should change it into:
```suggestion
  if ((touch_state & 0x87) == 0)
```

clydebarrow replied:
Bit 7 must be set for the data to be valid. The lower 3 bits encode the number of touches, which can range from 0-5. Your change would mean that an invalid byte of e.g. 0x01 would be incorrectly processed as a single touch. The code as I had it was based on reading the datasheet and actual testing. The check for the most significant bit is required, and should be reinstated.

nielsnl68 replied:
okay i understand it now :D thanks for explaining

clydebarrow commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -49,27 +47,22 @@ void GT911Touchscreen::setup() {
   i2c::ErrorCode err;
   touchscreen::TouchPoint tp;
   uint8_t touch_state = 0;
   uint8_t data[MAX_TOUCHES + 1][8];  // 8 bytes each for each point, plus extra space for the key byte
```
```suggestion
  uint8_t data[MAX_TOUCHES + 1][8];  // 8 for each point, plus extra space for the keypress info
```

clydebarrow replied:
CI objects to "byte" appearing even in comments!

jesserockz replied:
The actual issue is the regex that sees it like this:
```c++
  byte

  err = this->write(GET_TOUCH_STATE, sizeof(GET_TOUCH_STATE), false);
```

clydebarrow replied:
Ah regexes - we can't live without them but don't we hate them? 
So the regex should exclude comment lines.

jesserockz replied:
https://github.com/esphome/esphome/pull/5851

clydebarrow commented on esphome/components/ektf2232/touchscreen/ektf2232.cpp:
```diff
@@ -75,45 +66,15 @@ void EKTF2232Touchscreen::loop() {
-    this->defer([this, tp]() { this->send_touch_(tp); });
+    x_raw = (d[0] & 0xF0) << 4 | d[1];
+    y_raw = (d[0] & 0x0F) << 8 | d[2];
+    this->set_raw_touch_posistion_(i, x_raw, y_raw);
```
should be `set_raw_touch_position_`

nielsnl68 replied:
not sure where that "**S**" was coming from. :D

clydebarrow commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -49,27 +47,22 @@ void GT911Touchscreen::setup() {
-void GT911Touchscreen::loop() {
+void GT911Touchscreen::update_touches() {
   i2c::ErrorCode err;
   touchscreen::TouchPoint tp;
```
Is `tp` still needed?

jesserockz replied:
Nope 
```suggestion
```

jesserockz committed:
Apply suggestions from code review

Co-authored-by: Clyde Stubbs <2366188+clydebarrow@users.noreply.github.com>

jesserockz committed:
Spelling

jesserockz committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

jesserockz committed:
Remove unused config variables

jesserockz committed:
Change i type and remove commented code

jesserockz commented on esphome/components/ft63x6/ft63x6.cpp:
```diff
@@ -0,0 +1,99 @@
+    return;
+  }
+
+  uint8_t touch_id = read_touch_id_(FT63X6_ADDR_TOUCH1_ID);  // id1 = 0 or 1
+  int16_t x = read_touch_coordinate_(FT63X6_ADDR_TOUCH1_X);
+  int16_t y = read_touch_coordinate_(FT63X6_ADDR_TOUCH1_Y);
+  set_raw_touch_position_(touch_id, x, y);
```
```suggestion
  uint8_t touch_id = this->read_touch_id_(FT63X6_ADDR_TOUCH1_ID);  // id1 = 0 or 1
  int16_t x = this->read_touch_coordinate_(FT63X6_ADDR_TOUCH1_X);
  int16_t y = this->read_touch_coordinate_(FT63X6_ADDR_TOUCH1_Y);
  this->set_raw_touch_position_(touch_id, x, y);
```

jesserockz commented on esphome/components/ft63x6/ft63x6.cpp:
```diff
@@ -0,0 +1,99 @@
+  set_raw_touch_position_(touch_id, x, y);
+
+  if (touch_count >= 2) {
+    touch_id = read_touch_id_(FT63X6_ADDR_TOUCH2_ID);  // id2 = 0 or 1(~id1 & 0x01)
+    x = read_touch_coordinate_(FT63X6_ADDR_TOUCH2_X);
+    y = read_touch_coordinate_(FT63X6_ADDR_TOUCH2_Y);
+    set_raw_touch_position_(touch_id, x, y);
```
```suggestion
    touch_id = this->read_touch_id_(FT63X6_ADDR_TOUCH2_ID);  // id2 = 0 or 1(~id1 & 0x01)
    x = this->read_touch_coordinate_(FT63X6_ADDR_TOUCH2_X);
    y = this->read_touch_coordinate_(FT63X6_ADDR_TOUCH2_Y);
    this->set_raw_touch_position_(touch_id, x, y);
```

jesserockz commented on esphome/components/ft63x6/ft63x6.cpp:
```diff
@@ -0,0 +1,99 @@
+  LOG_PIN("  Reset Pin: ", this->reset_pin_);
+}
+
+uint8_t FT63X6Touchscreen::read_touch_count_() { return read_byte_(FT63X6_ADDR_TOUCH_COUNT); }
```
```suggestion
uint8_t FT63X6Touchscreen::read_touch_count_() { return this->read_byte_(FT63X6_ADDR_TOUCH_COUNT); }
```

jesserockz commented on esphome/components/ft63x6/ft63x6.cpp:
```diff
@@ -0,0 +1,99 @@
+// Touch functions
+uint16_t FT63X6Touchscreen::read_touch_coordinate_(uint8_t coordinate) {
+  uint8_t read_buf[2];
+  read_buf[0] = read_byte_(coordinate);
+  read_buf[1] = read_byte_(coordinate + 1);
```
```suggestion
  read_buf[0] = this->read_byte_(coordinate);
  read_buf[1] = this->read_byte_(coordinate + 1);
```

jesserockz commented on esphome/components/ft63x6/ft63x6.cpp:
```diff
@@ -0,0 +1,99 @@
+  read_buf[1] = read_byte_(coordinate + 1);
+  return ((read_buf[0] & 0x0f) << 8) | read_buf[1];
+}
+uint8_t FT63X6Touchscreen::read_touch_id_(uint8_t id_address) { return read_byte_(id_address) >> 4; }
```
```suggestion
uint8_t FT63X6Touchscreen::read_touch_id_(uint8_t id_address) { return this->read_byte_(id_address) >> 4; }
```

jesserockz commented on esphome/components/ft63x6/touchscreen.py:
```diff
@@ -0,0 +1,45 @@
+    await touchscreen.register_touchscreen(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_INTERRUPT_PIN in config:
+        interrupt_pin = await cg.gpio_pin_expression(config[CONF_INTERRUPT_PIN])
```
```suggestion
    if interrupt_pin_config := config.get(CONF_INTERRUPT_PIN):
        interrupt_pin = await cg.gpio_pin_expression(interrupt_pin_config)
```

jesserockz commented on esphome/components/ft63x6/touchscreen.py:
```diff
@@ -0,0 +1,45 @@
+    if CONF_INTERRUPT_PIN in config:
+        interrupt_pin = await cg.gpio_pin_expression(config[CONF_INTERRUPT_PIN])
+        cg.add(var.set_interrupt_pin(interrupt_pin))
+    if CONF_RESET_PIN in config:
+        reset_pin = await cg.gpio_pin_expression(config[CONF_RESET_PIN])
```
```suggestion
    if reset_pin_config := config.get(CONF_RESET_PIN):
        reset_pin = await cg.gpio_pin_expression(reset_pin_config)
```

jesserockz commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -80,29 +72,10 @@ void GT911Touchscreen::loop() {
-        break;
-    }
-    this->defer([this, tp]() { this->send_touch_(tp); });
+    set_raw_touch_position_(id, x, y);
```
```suggestion
    this->set_raw_touch_position_(id, x, y);
```

jesserockz commented on esphome/components/touchscreen/__init__.py:
```diff
@@ -3,44 +3,65 @@
+        cv.Optional(CONF_ON_RELEASE): automation.validate_automation(single=True),
     }
-)
+).extend(cv.polling_component_schema("50ms"))
```
If users dont need to be able to change this time and it is going to always poll at 50ms, it shouldnt be a PollingComponent, but should internally use `set_interval` and run the callback itself so users do not have the opportunity to change it.

nielsnl68 replied:
the idea of using the polling component is that people can use it when they have no interrupt pin enabled.

jesserockz commented on esphome/components/touchscreen/binary_sensor/touchscreen_binary_sensor.cpp:
```diff
@@ -10,15 +10,13 @@ void TouchscreenBinarySensor::setup() {
   bool touched = (tp.x >= this->x_min_ && tp.x <= this->x_max_ && tp.y >= this->y_min_ && tp.y <= this->y_max_);
-
-  if (this->page_ != nullptr) {
+  if ((this->page_ != nullptr) && (this->parent_->get_display() != nullptr)) {
```
The display cant be null with these changes.
```suggestion

  if (this->page_ != nullptr) {
```

jesserockz committed:
Discard changes to esphome/components/touchscreen/binary_sensor/touchscreen_binary_sensor.h

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+  touch = (z_raw >= this->threshold_);
+  if (touch) {
+    read_adc_(0xD1 /* X */);  // dummy Y measure, 1st is always noisy
+    data[0] = read_adc_(0x91 /* Y */);
+    data[1] = read_adc_(0xD1 /* X */);  // make 3 x-y measurements
+    data[2] = read_adc_(0x91 /* Y */);
+    data[3] = read_adc_(0xD1 /* X */);
+    data[4] = read_adc_(0x91 /* Y */);
```
```suggestion
    data[0] = this->read_adc_(0x91 /* Y */);
    data[1] = this->read_adc_(0xD1 /* X */);  // make 3 x-y measurements
    data[2] = this->read_adc_(0x91 /* Y */);
    data[3] = this->read_adc_(0xD1 /* X */);
    data[4] = this->read_adc_(0x91 /* Y */);
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+
+  enable();
+
+  int16_t touch_pressure_1 = read_adc_(0xB1 /* touch_pressure_1 */);
+  int16_t touch_pressure_2 = read_adc_(0xC1 /* touch_pressure_2 */);
```
```suggestion
  int16_t touch_pressure_1 = this->read_adc_(0xB1 /* touch_pressure_1 */);
  int16_t touch_pressure_2 = this->read_adc_(0xC1 /* touch_pressure_2 */);
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+    data[4] = read_adc_(0x91 /* Y */);
+  }
+
+  data[5] = read_adc_(0xD0 /* X */);  // Last X touch power down
```
```suggestion
  data[5] = this->read_adc_(0xD0 /* X */);  // Last X touch power down
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+
+    ESP_LOGV(TAG, "Touchscreen Update [%d, %d], z = %d", x_raw, y_raw, z_raw);
+
+    set_raw_touch_position_(0, x_raw, y_raw, z_raw);
```
```suggestion
    this->set_raw_touch_position_(0, x_raw, y_raw, z_raw);
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+int16_t XPT2046Component::read_adc_(uint8_t ctrl) {  // NOLINT
+  uint8_t data[2];
+
+  write_byte(ctrl);
```
```suggestion
  this->write_byte(ctrl);
```

jesserockz commented on esphome/components/xpt2046/touchscreen/xpt2046.cpp:
```diff
@@ -0,0 +1,113 @@
+
+  write_byte(ctrl);
+  delay(1);
+  data[0] = read_byte();
+  data[1] = read_byte();
```
```suggestion
  data[0] = this->read_byte();
  data[1] = this->read_byte();
```

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
impl Jesse's suggestions

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

nielsnl68 committed:
fix typo

nielsnl68 committed:
remove unneeded "component" class ref's

nielsnl68 committed:
oeps

nielsnl68 committed:
update touch validation

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

gaaf committed:
core: Don't yield when not delaying loop

On ESP32-IDF, yield is not necessary as FreeRTOS is preemptive.
On other archs there's no real need to yield just before the end of the loop.

nielsnl68 committed:
remove the rotation option in favor for mirror and
swap settings like the Display has.
Also added an **skip_update_** boolean.

nielsnl68 committed:
fix clang isues

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
add missing calibration settings

nielsnl68 committed:
remove the multitouch version of *touch* method.
from rhe `TouchListener` interface.

nielsnl68 committed:
`on_update` is now fired at all changes

nielsnl68 committed:
GT911: add skip_update when driver not ready

clydebarrow commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -37,40 +36,48 @@ void GT911Touchscreen::setup() {
+    uint8_t data[4];
+    err = this->write(GET_MAX_VALUES, 2);
+    if (err == i2c::ERROR_OK) {
+      err = this->read(&data, 1);
```
```suggestion
      err = this->read(data, sizeof(data));
```

clydebarrow commented on esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:
```diff
@@ -37,40 +36,48 @@ void GT911Touchscreen::setup() {
+    if (err == i2c::ERROR_OK) {
+      err = this->read(&data, 1);
+      if (err == i2c::ERROR_OK) {
+        this->x_raw_max_ = encode_uint16(data[0], data[1]);
+        this->y_raw_max_ = encode_uint16(data[2], data[3]);
```
```suggestion
        this->x_raw_max_ = encode_uint16(data[1], data[0]);
        this->y_raw_max_ = encode_uint16(data[3], data[2]);
```

clydebarrow committed:
Apply suggestions from code review

Fix gt911.

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
clang fix

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

nielsnl68 committed:
update test4.yaml

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
update test4.yaml restore some of the changes

nielsnl68 committed:
allow lambda's to get the current touchpoints.

nielsnl68 committed:
add test of touchscreen::get_touch() to test4.yaml

clydebarrow committed:
Update gt911 component to allow for no interrupt pin;
Touchscreen just needs a Display, not a DisplayBuffer.

clydebarrow committed:
Update gt911 component to allow for no interrupt pin;
Touchscreen just needs a Display, not a DisplayBuffer.

clydebarrow committed:
Rely on update() for polling.

clydebarrow committed:
Rely on update() for polling.

clydebarrow committed:
Disable update() calls if interrupts in use.

clydebarrow commented on esphome/components/touchscreen/touchscreen.cpp:
```diff
@@ -7,27 +7,125 @@ namespace touchscreen {
+  TouchPoint tp;
+  uint16_t x, y;
+  if (this->touches_.count(id) == 0) {
+    tp.state = 1;
```
It would make the code easier to read if these states were enum constants rather than plain numbers. Also a comment with a state machine description showing when the state changes would be helpful for the future.

clydebarrow commented on esphome/components/touchscreen/touchscreen.cpp:
```diff
@@ -7,27 +7,125 @@ namespace touchscreen {
+    std::swap(x, y);
+  }
+
+  tp.x = (uint16_t) ((int) x * this->get_width_() / 0xfff);
+  tp.y = (uint16_t) ((int) y * this->get_height_() / 0xfff);
```
```suggestion
  tp.x = (uint16_t) ((int) x * (this->get_width_() - 1) / 0xfff);
  tp.y = (uint16_t) ((int) y * (this->get_height_() - 1) / 0xfff);
```

clydebarrow replied:
Or alternatively, since the maximum value after normalisation will be 0xFFF:

Either way the resulting value should never == the width (or height) since the pixels are numbered 0:width-1.


```
tp.x = (uint16_t) ((int) x * this->get_width_() / 0x1000);
tp.y = (uint16_t) ((int) y * this->get_height_() / 0x1000);
```

nielsnl68 replied:
i like the later one better. One (two) calculation step less for the esp to calculated.

clydebarrow committed:
Add ft5x06 touchscreen

clydebarrow committed:
Add ft5x06 touchscreen

clydebarrow committed:
Fix height/width

clydebarrow committed:
Fix height/width

nielsnl68 committed:
fix clang and added @clydebarrow suggestion

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

nielsnl68 committed:
Make state use const values and update some var names

nielsnl68 committed:
fix clang

nielsnl68 committed:
oeps

nielsnl68 committed:
oeps

clydebarrow committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into nvds-touchscreen-optdisp

clydebarrow committed:
CODEOWNERS

clydebarrow committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into xxx

clydebarrow commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,54 +1,129 @@
+
+  optional<TouchPoint> get_touch() {
+    if (!this->touches_.empty()) {
+      for (auto i : this->touches_) {
+        return i.second;
+      }
```
```suggestion
      return this->touches_.begin().second;
```

clydebarrow commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,54 +1,129 @@
+    return {};
+  }
+
+  optional<TouchPoints_t> get_touches() {
```
Can't this one just return an empty vector if there are no touches, rather than wrapping it in an `optional`?

nielsnl68 committed:
Update esphome/components/touchscreen/touchscreen.h

Co-authored-by: Clyde Stubbs <2366188+clydebarrow@users.noreply.github.com>

nielsnl68 committed:
Make sure that the single touch is always using the right tp

clydebarrow committed:
Add Display as inherited class for inplate6 and ili9xxxx

clydebarrow committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into touchscreen

clydebarrow committed:
CODEOWNERS

clydebarrow committed:
Pointers need -> not .

gaaf committed:
core: Force minimum 1 ms delay in loop()

Workaround for giving lower priority tasks at least a timeslice.

On ESP320-IDF, some esphome components are creating tasks at prio 0. This
is probably wrong as that is the lowest prio (same as the idle task).

Even though it is highly unlikely the loop task (at prio 1) won't block during
its execution, let's try to prevent starvation of them (until they get fixed)
by sleeping a minimum time in the loop() task. This should give them at least
1 time slice per iteration.

This commit should probably be reverted after it is established the lower prio
tasks don't actually need any cpu time, or they get fixed to an appropriate
priority.

This is not an issue on non-FreeRTOS architectures as those run other
(housekeeping) tasks in-between loop() calls.

gaaf committed:
core: Don't account sleeping in loop() to the next loop iteration

The delay really is to extend the current loop. Don't account it to the next
loop. And account logging the config to the current loop.

This results in less jitter in the loop start times.

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

clydebarrow commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,54 +1,122 @@
+
+using TouchPoints_t = std::vector<TouchPoint>;
+
+struct TouchscreenInterupt {
```
Should be `TouchScreenInterrupt`

clydebarrow commented on esphome/components/touchscreen/touchscreen.h:
```diff
@@ -1,54 +1,122 @@
 namespace esphome {
 namespace touchscreen {
 
+static const uint8_t STATE_RELAESED = 0x00;
```
Should be `STATE_RELEASED`

nielsnl68 committed:
suggestion from Clyde

clydebarrow committed:
Merge commit 'refs/pull/5869/head' of https://github.com/esphome/esphome into touchscreen

nielsnl68 committed:
Merge branch 'dev' into nvds-touchscreen-optdisp

clydebarrow committed:
Merge branch 'nvds-touchscreen-optdisp' of https://github.com/nielsnl68/esphome into touchscreen

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into touchscreen

clydebarrow committed:
Fix spelling errors

clydebarrow committed:
Revert incorrect merge

clydebarrow committed:
Merge branch 'dev' into nvds-touchscreen-optdisp
